src/cpu/simple/probes/looppoint_analysis.cc

```diff
@@ -0,0 +1,280 @@
+            lpaManager->incrementGlobalInstCounter();
+        }
+        else if (lpaManager->ifValidControl(pc))
+        {
+            // if it is categorized as a valid control instruction
+            bbInstCounter ++;
+            lpaManager->incrementGlobalInstCounter();
+            lpaManager->updateBBInstMap(pc, bbInstCounter);
+            updateLocalBBV(pc);
+            lpaManager->updateGlobalBBV(pc);
+            bbInstCounter = 0;
+            if (lpaManager->ifBackwardBranch(pc))
+            {
+                // if it is categorized as a backward branch
+                lpaManager->countBackwardBranch(pc);
+            }
+        }
+        return;
+    }
+
+    lpaManager->updateEncountered(pc);
+    // if we have not encountered this pc before, we should now update it to
+    // the corresponding category
+
+    if (!thread->getIsaPtr()->inUserMode())
+    {
+        // ignore this if it is not in user mode
+        return;
+    }
+
+    if (bbValidAddrRange.end() > 0 &&
+        (pc < bbValidAddrRange.start() || pc > bbValidAddrRange.end()))
+    {
+        // ignore this if it is not in the valid address range
+        return;
+    }
+
+    if (bbExcludedAddrRanges.size() > 0)
+    {
+        for (int i = 0; i < bbExcludedAddrRanges.size(); i++)
+        {
+            if (pc >= bbExcludedAddrRanges[i].start() &&
+                pc <= bbExcludedAddrRanges[i].end())
+            {
+                // ignore this if it is in the excluded address range
+                return;
+            }
+        }
+    }
+
+    bbInstCounter++;
+    lpaManager->incrementGlobalInstCounter();
+
+    if (inst->isControl())
+    {
+        // if it is a control instruction, we see it as the end of a basic
+        // block
+        lpaManager->updateValidControl(pc);
+        lpaManager->updateBBInstMap(pc, bbInstCounter);
+        updateLocalBBV(pc);
+        lpaManager->updateGlobalBBV(pc);
+        bbInstCounter = 0;
+
+        if (markerValidAddrRange.end() > 0 &&
+         (pc < markerValidAddrRange.start() || pc > markerValidAddrRange.end())
+        )
+        {
+            // if it is not in the marker valid address range, we do not
+            // consider it as a possible marker used loop branch instruction
+            return;
+        }
+
+        if (inst->isDirectCtrl())
+        {
+            // We only consider direct control instructions as possible
+            // loop branch instructions because it is PC-relative and it
+            // excludes return instructions.
+            if (pcstate.npc() < pc)
+            {
+                lpaManager->updateBackwardBranch(pc);
+                lpaManager->countBackwardBranch(pc);
+            }
+
+        }
+    }
+    else
+    {
+        lpaManager->updateValidNotControl(pc);
+    }
+}
+
+void
+LooppointAnalysis::regProbeListeners()
+{
+    if (ifListening)
+    {
+        if (listeners.size() == 0) {
+            listeners.push_back(new looppointAnalysisListener(this,
```

Comment:
See #1439 to make sure this doesn't conflict

Reply:
It does conflict because I am using the constructor of the ProbeListener object to add the listeners to the probe point and remove them using its destructor too; but I think I can push a patch to fix it later after we decide what is the best way to add the listeners in that pull request.